docs/configuration.md
# configuration


#### configs


config | type | valid values | has default | default value | environment variable | description
--- | --- | --- | --- | --- | --- | ---
node.network | string | ['mainnet', 'testnet', 'regtest', 'simnet'] | yes | "testnet" | SQUEAKNODE_NODE_NETWORK | Which network to use.
node.price_msat | int | [0,...] | yes | 10000 | SQUEAKNODE_NODE_PRICE_MSAT | The price to sell squeaks to other peers in millisatoshis.
node.max_squeaks | int | [0,...] | yes | 10000 | SQUEAKNODE_NODE_MAX_SQUEAKS | The absolute maximum number of squeaks allowed in the database.
node.max_squeaks_per_public_key_per_block | int | [0,...] | yes | 1000 | SQUEAKNODE_NODE_MAX_SQUEAKS_PER_PUBLIC_KEY_PER_BLOCK | The maximum number of squeaks for an any public key with any block height.
node.sqrd_dir_path | string | | yes | "<USER_HOME>/.sqk" | SQUEAKNODE_NODE_SQRD_DIR_PATH | The directory to store application data (only if using sqlite as database backend).
node.log_level | string | | yes | "INFO" | SQUEAKNODE_NODE_LOG_LEVEL | The log level to use.
node.sent_offer_retention_s | int | [0,...] | yes | 86400 | SQUEAKNODE_NODE_SENT_OFFER_RETENTION_S | The amount of time in seconds to keep a sent offer after expiry before deleting it.
node.received_offer_retention_s | int | [0,...] | yes | 86400 | SQUEAKNODE_NODE_RECEIVED_OFFER_RETENTION_S | The amount of time in seconds to keep a received offer after download before deleting it.
node.subscribe_invoices_retry_s | int | [0,...] | yes | 10 | SQUEAKNODE_NODE_SUBSCRIBE_INVOICES_RETRY_S | The amount of time in seconds to wait after a subscription failure to retry subscribing settled invoices.
node.squeak_retention_s | int | [0,...] | yes | 604800 | SQUEAKNODE_NODE_SQUEAK_RETENTION_S | The amount of time in seconds to keep a squeak after download before deleting it. This only applies to squeaks that are not liked or created by a signing profile.
node.squeak_deletion_interval_s | int | [0,...] | yes | 10 | SQUEAKNODE_NODE_SQUEAK_DELETION_INTERVAL_S | The amount of time in seconds to wait in between deleting old squeaks.
node.offer_deletion_interval_s | int | [0,...] | yes | 10 | SQUEAKNODE_NODE_OFFER_DELETION_INTERVAL_S | The amount of time in seconds to wait in between deleting old offers.
node.interest_block_interval | int | [0,...] | yes | 2016 | SQUEAKNODE_NODE_INTEREST_BLOCK_INTERVAL | The number of blocks (starting from the most recent and descending) that this node will attempt to find squeaks with matching block height.
node.peer_autoconnect_interval_s | int | [0,...] | yes | 10 | SQUEAKNODE_NODE_PEER_AUTOCONNECT_INTERVAL_S | The amount of time in seconds to wait in between trying to connect autoconnect peers.
bitcoin.rpc_host | string | | yes | "localhost" | SQUEAKNODE_BITCOIN_RPC_HOST | The host of the bitcoin node to connect.
bitcoin.rpc_port | int | | yes | 18334 | SQUEAKNODE_BITCOIN_RPC_HOST | The port of the bitcoin node to connect.
bitcoin.rpc_user | string | | yes | "" | SQUEAKNODE_BITCOIN_RPC_USER | The username to use for authentication on the bitcoin node.
bitcoin.rpc_pass | string | | yes | "" | SQUEAKNODE_BITCOIN_RPC_PASS | The password to use for authentication on the bitcoin node.
bitcoin.rpc_use_ssl | boolean | [true, false] | yes | false | SQUEAKNODE_BITCOIN_USE_SSL | Use SSL for the connection to the bitcoin node.
bitcoin.rpc_ssl_cert | str |  | yes | "" | SQUEAKNODE_BITCOIN_SSL_CERT | The path to the SSL cert to use for connection to the bitcoin node, if one is used.
lightning.backend | string | | yes | "lnd" | SQUEAKNODE_LIGHTNING_BACKEND | The Lightning Network node implementation to use.
lightning.external_host | string | | yes | "" | SQUEAKNODE_LIGHTNING_EXTERNAL_HOST | The host of the lightning node to share with other peers.
lightning.external_port | int | | yes | | SQUEAKNODE_LIGHTNING_EXTERNAL_PORT | The port of the lightning node to share with other peers.
lightning.lnd_rpc_host | string | | yes | "localhost" | SQUEAKNODE_LIGHTNING_LND_RPC_HOST | The host of the LND node to connect.
lightning.lnd_rpc_port | int | | yes | | SQUEAKNODE_LIGHTNING_LND_RPC_PORT | The port of the LND node to use for RPC connections.
lightning.lnd_tls_cert_path | string | | yes | "" | SQUEAKNODE_LIGHTNING_LND_TLS_CERT_PATH | The path to the TLS certificate to use for LND connection.
lightning.lnd_macaroon_path | string | | yes | "" | SQUEAKNODE_LIGHTNING_LND_MACAROON_PATH | The path to the macaroon to use for LND connection.
lightning.clightning_rpc_file | string | | yes | "" | SQUEAKNODE_LIGHTNING_CLIGHTNING_RPC_FILE | The path to the file to use for c-lightning RPC.
tor.proxy_ip | string | | yes | "" | SQUEAKNODE_TOR_PROXY_IP | The ip address or host of the SOCKS5 Tor proxy, if one is used.
tor.proxy_port | int | | yes | 0 | SQUEAKNODE_TOR_PROXY_PORT | The port of the SOCKS5 Tor proxy, is one is used.
db.connection_string | string | | yes | "" | SQUEAKNODE_DB_CONNECTION_STRING | The connection string to use to connect to a SQL database. If none is specified, a sqlite database will be used on the local file system.
rpc.enabled | boolean | [true, false] | yes | false | SQUEAKNODE_RPC_ENABLED | Accept RPC commands or not.
rpc.host | string | | yes | "0.0.0.0" | SQUEAKNODE_RPC_HOST | Host to listen for rpc connections.
rpc.port | int | | yes | 8994 | SQUEAKNODE_RPC_PORT | Port to listen for rpc connections.
webadmin.enabled | boolean | [true, false] | yes | false | SQUEAKNODE_WEBADMIN_ENABLED | Run a web admin server or not.
webadmin.host | string | | yes | "0.0.0.0" | SQUEAKNODE_WEBADMIN_HOST | Host to user for serving admin web server.
webadmin.port | int | | yes | 12994 | SQUEAKNODE_WEBADMIN_PORT | Port to user for serving admin web server.
webadmin.username | string | | yes | "" | SQUEAKNODE_WEBADMIN_USERNAME | Username to sign in to admin web server.
webadmin.password | string | | yes | "" | SQUEAKNODE_WEBADMIN_PASSWORD | Password to sign in to admin web server.
webadmin.use_ssl | boolean | | yes | false | SQUEAKNODE_WEBADMIN_USE_SSL | Use SSL for admin web server or not.
webadmin.login_disabled | boolean | | yes | false | SQUEAKNODE_WEBADMIN_LOGIN_DISABLED | Disable requiring login for web server or not.
webadmin.allow_cors | boolean | | yes | false | SQUEAKNODE_WEBADMIN_ALLOW_CORS | Allow CORS requests to admin web server or not.
server.enabled | boolean | | yes | true | SQUEAKNODE_SERVER_ENABLED | If true, then accept inbound connections from other peers.
server.host | string | | yes | "0.0.0.0" | SQUEAKNODE_SERVER_HOST | Host to user for accepting inbound peer connections.
server.port | int | | yes | 8555/18555 | SQUEAKNODE_SERVER_PORT | Port to user for accepting inbound peer connections.
server.external_address | string | | yes | "" | SQUEAKNODE_SERVER_EXTERNAL_ADDRESS | The address that other nodes should use to open a connection to this node.


CHANGELOG.md
0.1.165 - 2021-11-04
===================

### Features
* Add twitter forwarder by @yzernik in https://github.com/yzernik/squeaknode/pull/1755
* Fix button text for submit button in add twitter account dialog by @yzernik in https://github.com/yzernik/squeaknode/pull/1757
* Make rpc enabled config set to false by default by @yzernik in https://github.com/yzernik/squeaknode/pull/1758
* Add custom config for autoconnect peer interval by @yzernik in https://github.com/yzernik/squeaknode/pull/1759

0.1.164 - 2021-11-02
===================

### Features
* Include network name in profile dropdown in header by @yzernik in https://github.com/yzernik/squeaknode/pull/1743
* Add license section to readme by @yzernik in https://github.com/yzernik/squeaknode/pull/1747
* Make rpc enabled by default by @yzernik in https://github.com/yzernik/squeaknode/pull/1749
* Update received offers in display even when none downloaded by @yzernik in https://github.com/yzernik/squeaknode/pull/1750

0.1.163 - 2021-10-28
===================

### Features
* Make download async by @yzernik in https://github.com/yzernik/squeaknode/pull/1725
* Add download in progress dialog @yzernik in https://github.com/yzernik/squeaknode/pull/1732
* Remove hash from title of buy squeak dialog by @yzernik in https://github.com/yzernik/squeaknode/pull/1741

0.1.162 - 2021-10-24
===================

### Features
* Add tests for db queries
* Make received offer retention a config by @yzernik in https://github.com/yzernik/squeaknode/pull/1710
* Move external address config to server section by @yzernik in https://github.com/yzernik/squeaknode/pull/1712
* Rename configs in server section by @yzernik in https://github.com/yzernik/squeaknode/pull/1713
* Update configuration doc for server configs by @yzernik in https://github.com/yzernik/squeaknode/pull/1714
* Rename rpc config section by @yzernik in https://github.com/yzernik/squeaknode/pull/1716
* Make update subscriptions async by @yzernik in https://github.com/yzernik/squeaknode/pull/1723

0.1.161 - 2021-10-14
===================

### Features
* Add more tests
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1525
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1526
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1527
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1530
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1533
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1541
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1543
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1548
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1560
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1562
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1564
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1578
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1579
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1580
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1583
  -  by @yzernik in https://github.com/yzernik/squeaknode/pull/1586
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1587
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1588
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1590
  - by @yzernik in https://github.com/yzernik/squeaknode/pull/1591
* Use deserialized block header in info object by @yzernik in https://github.com/yzernik/squeaknode/pull/1536
* Use decorator for protobuf flask routes by @yzernik in https://github.com/yzernik/squeaknode/pull/1538
* Create lightning client base class by @yzernik in https://github.com/yzernik/squeaknode/pull/1542
* Add toggle for connect with tor by @yzernik in https://github.com/yzernik/squeaknode/pull/1550
* Add rpc to get default peer port by @yzernik in https://github.com/yzernik/squeaknode/pull/1556
* Rename core crypto module by @yzernik in https://github.com/yzernik/squeaknode/pull/1563
* Remove secret key field of sent offer by @yzernik in https://github.com/yzernik/squeaknode/pull/1568
* Add check for payment point in unpack offer by @yzernik in https://github.com/yzernik/squeaknode/pull/1577

0.1.160 - 2021-10-07
===================

### Features
- Add unit test for timeline query.
    - #1518
- Add OpenTimestamps github action.
    - #1524

0.1.159 - 2021-10-04
===================

### Features
- Improve handling of duplicate entries in database.
    - #1478
- Refactor core functions to separate modules
    - #1460, #1464, #1468, #1473

0.1.158 - 2021-10-01
===================

### Features
- Show full date and time of squeak in tooltip.
    - #1393 PR by @abhiShandy

0.1.157 - 2021-09-28
===================

### Fixes
- Upgrade SQLAlchemy to version 1.4.
    - #1438

0.1.156 - 2021-09-27
===================

### Fixes
- Fix display external address text with full width.
    - #1429 PR by @abhiShandy

0.1.155 - 2021-09-27
===================

### Features
- Make p2p protocol more efficient by only requesting
  incremental changes when interests are updated.
    - #1424 and #1423

0.1.154 - 2021-09-24
===================

### Features
- Include saved peer name in list of connected peers.
    - #1361 and #1362
- Add button in UI to convert connected peer to saved peer.
    - #1365
- Improve dialog for showing external address, and include copy to
  clipboard.
    - #1380 PR by @abhiShandy
- Show connection status next to each item in saved peers list.
    - #1399

### Fixes
- Fix bug in p2p connection where the connection does not shut down
  properly in some cases.
    - #1397
- Fail gracefully in peer connection when LND is not available.
    - #1389 and #1395

0.1.153 - 2021-09-19
===================

### Features
- Better p2p networking. Inactive peer connection is now detected with
  `ping` message and stopped.
    - #1342 and #1347
- Better peer connection UI. `ConnectPeer` RPC is now synchronous, and

squeaknode/config/config.py
# MIT License
#
# Copyright (c) 2020 Jonathan Zernik
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
import logging
from pathlib import Path

from typedconfig import Config
from typedconfig import group_key
from typedconfig import key
from typedconfig import section
from typedconfig.source import DictConfigSource
from typedconfig.source import EnvironmentConfigSource
from typedconfig.source import IniFileConfigSource


logger = logging.getLogger(__name__)


DEFAULT_NETWORK = "testnet"
DEFAULT_PRICE_MSAT = 10000
DEFAULT_LOG_LEVEL = "INFO"
DEFAULT_MAX_SQUEAKS = 10000
DEFAULT_MAX_SQUEAKS_PER_PUBLIC_KEY_PER_BLOCK = 100
DEFAULT_SERVER_RPC_HOST = "0.0.0.0"
DEFAULT_SERVER_RPC_PORT = 8555
DEFAULT_EXTERNAL_PORT = 8555
DEFAULT_ADMIN_RPC_HOST = "0.0.0.0"
DEFAULT_ADMIN_RPC_PORT = 8994
DEFAULT_WEBADMIN_HOST = "0.0.0.0"
DEFAULT_WEBADMIN_PORT = 12994
DEFAULT_BITCOIN_RPC_HOST = "localhost"
DEFAULT_BITCOIN_RPC_PORT = 18334
BITCOIN_RPC_PORT = {
    "mainnet": 8332,
    "testnet": 18332,
    "simnet": 18556,
}
DEFAULT_LIGHTNING_BACKEND = "lnd"
DEFAULT_LND_PORT = 9735
DEFAULT_LND_RPC_PORT = 10009
DEFAULT_CLIGHTNING_DIR = ".lightning"
DEFAULT_CLIGHTNING_RPC_FILE = str(
    Path.home() / DEFAULT_CLIGHTNING_DIR / 'lightning-rpc')
DEFAULT_SQRD_DIR = ".sqrd"
DEFAULT_SQRD_DIR_PATH = str(Path.home() / DEFAULT_SQRD_DIR)
DEFAULT_LND_RPC_HOST = "localhost"
DEFAULT_INTEREST_BLOCK_INTERVAL = 2016
DEFAULT_SENT_OFFER_RETENTION_S = 86400
DEFAULT_RECEIVED_OFFER_RETENTION_S = 86400
DEFAULT_OFFER_DELETION_INTERVAL_S = 10
DEFAULT_PEER_DOWNLOAD_INTERVAL_S = 30
DEFAULT_SUBSCRIBE_INVOICES_RETRY_S = 10
DEFAULT_SQUEAK_RETENTION_S = 604800
DEFAULT_SQUEAK_DELETION_INTERVAL_S = 10
DEFAULT_FORWARD_TWEETS_RETRY_S = 10


@section('bitcoin')
class BitcoinConfig(Config):
    rpc_host = key(cast=str, required=False, default=DEFAULT_BITCOIN_RPC_HOST)
    rpc_port = key(cast=int, required=False, default=DEFAULT_BITCOIN_RPC_PORT)
    rpc_user = key(cast=str, required=False, default="")
    rpc_pass = key(cast=str, required=False, default="")
    rpc_use_ssl = key(cast=bool, required=False, default=False)
    rpc_ssl_cert = key(cast=str, required=False, default="")


@section('lightning')
class LightningConfig(Config):
    backend = key(cast=str, required=False, default=DEFAULT_LIGHTNING_BACKEND)
    external_host = key(cast=str, required=False, default="")
    external_port = key(cast=int, required=False, default=DEFAULT_LND_PORT)
    lnd_rpc_host = key(cast=str, required=False, default=DEFAULT_LND_RPC_HOST)
    lnd_rpc_port = key(cast=int, required=False, default=DEFAULT_LND_RPC_PORT)
    lnd_tls_cert_path = key(cast=str, required=False, default="")
    lnd_macaroon_path = key(cast=str, required=False, default="")
    clightning_rpc_file = key(cast=str, required=False,
                              default=DEFAULT_CLIGHTNING_RPC_FILE)


@section('tor')
class TorConfig(Config):
    proxy_ip = key(cast=str, required=False, default="")
    proxy_port = key(cast=int, required=False, default=0)


@section('webadmin')
class WebadminConfig(Config):
    enabled = key(cast=bool, required=False, default=False)
    host = key(cast=str, required=False, default=DEFAULT_WEBADMIN_HOST)
    port = key(cast=int, required=False, default=DEFAULT_WEBADMIN_PORT)
    username = key(cast=str, required=False, default="")
    password = key(cast=str, required=False, default="")
    use_ssl = key(cast=bool, required=False, default=False)
    login_disabled = key(cast=bool, required=False, default=False)
    allow_cors = key(cast=bool, required=False, default=False)


@section('node')
class NodeConfig(Config):
    network = key(
        cast=str, required=False, default=DEFAULT_NETWORK)
    price_msat = key(
        cast=int, required=False, default=DEFAULT_PRICE_MSAT)
    max_squeaks = key(
        cast=int, required=False, default=DEFAULT_MAX_SQUEAKS)
    max_squeaks_per_public_key_per_block = key(
        cast=int, required=False, default=DEFAULT_MAX_SQUEAKS_PER_PUBLIC_KEY_PER_BLOCK)
    sqrd_dir_path = key(
        cast=str, required=False, default=DEFAULT_SQRD_DIR_PATH)
    log_level = key(
        cast=str, required=False, default=DEFAULT_LOG_LEVEL)
    sent_offer_retention_s = key(
        cast=int, required=False, default=DEFAULT_SENT_OFFER_RETENTION_S)
    received_offer_retention_s = key(
        cast=int, required=False, default=DEFAULT_RECEIVED_OFFER_RETENTION_S)
    subscribe_invoices_retry_s = key(
        cast=int, required=False, default=DEFAULT_SUBSCRIBE_INVOICES_RETRY_S)
    squeak_retention_s = key(
        cast=int, required=False, default=DEFAULT_SQUEAK_RETENTION_S)
    squeak_deletion_interval_s = key(
        cast=int, required=False, default=DEFAULT_SQUEAK_DELETION_INTERVAL_S)
    offer_deletion_interval_s = key(
        cast=int, required=False, default=DEFAULT_OFFER_DELETION_INTERVAL_S)
    interest_block_interval = key(
        cast=int, required=False, default=DEFAULT_INTEREST_BLOCK_INTERVAL)
    peer_download_interval_s = key(
        cast=int, required=False, default=DEFAULT_PEER_DOWNLOAD_INTERVAL_S)


@section('db')
class DbConfig(Config):
    connection_string = key(cast=str, required=False, default="")


@section('twitter')
class TwitterConfig(Config):
    forward_tweets_retry_s = key(
        cast=int, required=False, default=DEFAULT_FORWARD_TWEETS_RETRY_S)


class SqueaknodeConfig(Config):
    bitcoin = group_key(BitcoinConfig)
    lightning = group_key(LightningConfig)
    tor = group_key(TorConfig)
    webadmin = group_key(WebadminConfig)
    node = group_key(NodeConfig)
    db = group_key(DbConfig)
    twitter = group_key(TwitterConfig)
    # description = key(cast=str, section_name="general")

    def __init__(self, config_path=None, dict_config=None):
        super().__init__()
        self.prefix = "SQUEAKNODE"
        self.config_path = config_path
        self.dict_config = dict_config

    def read(self):
        if self.dict_config is not None:
            self.add_source(DictConfigSource(self.dict_config))
        self.add_source(EnvironmentConfigSource(prefix=self.prefix))
        if self.config_path is not None:
            self.add_source(IniFileConfigSource(self.config_path))
        return super().read()
